ecosystem/sep-0010.md

```diff
@@ -18,12 +18,21 @@ This SEP defines the standard way for clients such as wallets or exchanges to cr
 
 This protocol is a variation of mutual challenge-response, which uses Stellar transactions to encode challenges and responses.
 
-The authentication flow is as follows:
+It involves three parties:
+- A home domain hosting a [SEP-1 stellar.toml](sep-0001.md) containing a `WEB_AUTH_ENDPOINT` value.
+- A web auth endpoint providing the GET and POST operations discussed in this document.
+- A client who will authenticate with a Stellar account.
+
+The discovery flow is as follows:
+
+1. The client retrieves the [SEP-1 stellar.toml](sep-0001.md) from the home domain. The home domain hosts the stellar.toml.
+1. The client looks up the `WEB_AUTH_ENDPOINT` value from the [SEP-1 stellar.toml](sep-0001.md).
 
-1. The client obtains a unique [`challenge`](#challenge), which is represented as specially formed Stellar transaction
+The authentication flow is as follows:
+1. The client obtains a unique [`challenge`](#challenge) from the `WEB_AUTH_ENDPOINT`, which is represented as specially formed Stellar transaction
 1. The client verifies that the transaction has an invalid sequence number 0.  This is extremely important to ensure the transaction isn't malicious.
 1. The client verifies that the transaction is signed by the `SIGNING_KEY` specified by the requested service's [SEP-1 stellar.toml](sep-0001.md).
-1. The client verifies that the transaction has a single Manage Data operation with its source account set to the user's account and value set to a nonce value. The client ignores the home domain included.
+1. The client verifies that the transaction has a single Manage Data operation with its source account set to the user's account and value set to a nonce value. The client verifies the home domain included is the home domain from the discovery flow.
```

Comment:
In SEP-10 v2.1.0 the operations were relaxed to allow any number of operations greater than one, so I think this is already achieved. See https://github.com/stellar/stellar-protocol/commit/5acf11be9837969ec883c2f9fd7cd3201728ff07.

Let me know what I'm missing if this isn't already addressed sufficiently in v2.1.0.

Reply:
@leighmcculloch Apparently I got confused but that might indicate that there's a room for improvement here. The way I read it was like two separate predicates on transaction and operation: 

> The client verifies that the transaction has a single Manage Data operation. This operation has a source account set to... => `tx.operations.size == 1 & valid_auth_op?(tx.operations[0])`.

And I guess what it suppose to mean is actually a single predicate

> The client verifies that `exactly one operation in the transaction is a Manage Data operation with its source account set to...`" => `tx.operations.find {|op| valid_auth_op?(op)}.size == 1`

Still I don't think it is clear how the client should proceed with the verification, it looks like it is expected to perform a search through transaction operations (which by itself is an extra complication, I think we can safely require that the auth operation is the first one), but the search condition a bit is misleading. Reading the current text exactly, it appears the client has to find operations with `source_account == account && value == nonce`, ensure that exactly one operation is found, and then verify that this operation contains correct home domain. But client doesn't know anything about the nonce and should only care that it exists, and the correct predicate is in fact `source_account == account && name == "#{home_domain} auth"`. Overall, it doesn't seem like the spec in the current edition can be directly translated to the code.

Does that makes sense, or am I missing something?

PS: Just realised that in the change suggested above I forgot about ` auth` suffix, so I updated my comment.

Reply:
Another point I forgot to mention is about the next step in the flow, which requires other operations in the transaction to have a) Manage Data type; and b) have the signer set to the server acc. It is now clear that this condition is necessary because of the "find the only operation which has user as source account" semantics introduced in 2.1.0, but if we switch to the approach I'm suggesting "first operation in the transaction must be the auth op", then both those restrictions on the extra operations can be lifted, which I think is great from future extensibility standpoint. Existing clients would only care about the first op in the transactions, and just ignore the rest, if there're any.

Reply:
I agree, we can do some things to improve the steps as described in the abstract, although this behavior is not something changing in v3 and is something described by v2.1.0 so we should improve it independently of the v2 vs v3. All SDKs I've reviewed implement the auth op as the first op as that is how the operation is represented further down in the SEP where the first op is always the auth op and zero or more ops are allowed to follow:
https://github.com/stellar/stellar-protocol/blob/a235c6366e0eef9c3f807222b1d3a8abad10405e/ecosystem/sep-0010.md#L112-L116
I think we could resolve this gap in a separate PR making it clear in the abstract that the auth op is the first operation.

Reply:
I am addressing this issue separately in https://github.com/stellar/stellar-protocol/pull/754.